### PR TITLE
Shrink unicode case-mapping LUTs by 24k

### PR DESCRIPTION
**Pull request by martingms**
I was looking into the binary bloat of a small program using `str::to_lowercase` and `str::to_uppercase`, and noticed that the lookup tables used for case mapping had a lot of zero-bytes in them. The reason for this is that since some characters map to up to three other characters when lower or uppercased, the LUTs store a `[char; 3]` for each character. However, the vast majority of cases only map to a single new character, in other words most of the entries are e.g. `(lowerc, [upperc, '\0', '\0'])`.
This PR introduces a new encoding scheme for these tables.

The changes reduces the size of my test binary by about 24K.

I've also done some `#[bench]`marks on unicode-heavy test data, and found that the performance of both `str::to_lowercase` and `str::to_uppercase` improves by up to 20%. These measurements are obviously very dependent on the character distribution of the data.

Someone else will have to decide whether this more complex scheme is worth it or not, I was just goofing around a bit and here's what came out of it :man_shrugging: No hard feelings if this isn't wanted!

### CODE REVIEW

**Comment by workingjubilee** [comment]
cc @thomcc

**Comment by nikic** [comment]
It would be possible to avoid the second table lookup by storing an index into the table with the long expansions. Something like `(1 << 24) | large_index`. This does mean that it can't be stored as `char` and needs to use `u32` instead, because these would be stored in the invalid code point space.

**Comment by martingms** [comment]
> It would be possible to avoid the second table lookup by storing an index into the table with the long expansions. Something like `(1 << 24) | large_index`. This does mean that it can't be stored as `char` and needs to use `u32` instead, because these would be stored in the invalid code point space.

That is a very good idea!

I'll see if I can get something working a bit later.

**Comment by Mark-Simulacrum** [comment]
r=me presuming perf looks ok

File: library/core/src/unicode/unicode_data.rs [hunk 1]
```diff
@@ -580,1815 +580,804 @@ pub mod white_space {
 
 #[rustfmt::skip]
 pub mod conversions {
+    const INDEX_MASK: u32 = 4194304;
```
Comment (scottmcm) [inline library/core/src/unicode/unicode_data.rs:583]: nit: might be clearer if this were `0x40_0000` instead, at least for me who only knows my powers of two up so high.

Comment (martingms) [reply library/core/src/unicode/unicode_data.rs]: I agree, pushed a fix to `0x400000` at least (dunno if there's some fmt operator that inserts `_`s?).

Comment (Mark-Simulacrum) [reply library/core/src/unicode/unicode_data.rs]: Ultimately this is the generated code, so I'm not too worried about readability in it.

**Comment by martingms** [comment]
> broken_heart Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4508023345/jobs/7936961374)

Seems unrelated?

**Pull request merged**